frontend/catalyst/jax_primitives.py

```diff
@@ -368,6 +403,133 @@ def _print_lowering(jax_ctx: mlir.LoweringRuleContext, *args, string=None, memre
+
+        # The transform.named_sequence needs a terminator called "transform.yield"
+        with ir.InsertionPoint(bb):
+            transform_yield_op = TransformYieldOp(operands_=[])  # pylint: disable=unused-variable
+
+    return named_sequence_op.results
+
+
+#
+# apply_registered_pass
+#
+@apply_registered_pass_p.def_abstract_eval
+def _apply_registered_pass_abstract_eval(*args, pass_name, options=None):
+    return AbstractTransformFunc()
+
+
+@apply_registered_pass_p.def_impl
+def _apply_registered_pass_def_impl(*args, pass_name, options=None):  # pragma: no cover
+    raise NotImplementedError()
+
+
+def _apply_registered_pass_lowering(
+    jax_ctx: mlir.LoweringRuleContext, *args, pass_name, options=None
+):
+    module = jax_ctx.module_context.module
+    named_sequence_op = None
+    for i in reversed(range(len(module.body.operations))):
+        # transform.named_sequence usually is at the end of the module, so look for it from the end
+        if module.body.operations[i].operation.name == "transform.named_sequence":
+            named_sequence_op = module.body.operations[i].operation
+            break
+    if named_sequence_op is None:
+        raise RuntimeError(
+            "transform.apply_registered_pass must be placed in a transform.named_sequence!"
+        )
+
+    # If there already is a apply_registered_pass, insert after the existing one
+    # Note that ir.InsertionPoint(op) sets the insertion point to immediately BEFORE the op
+    named_sequence_op_block = named_sequence_op.regions[0].blocks[0]
+    first_op_in_block = named_sequence_op_block.operations[0].operation
+
+    if first_op_in_block.name == "transform.apply_registered_pass":
+        _ = len(named_sequence_op_block.operations)
+        yield_op = named_sequence_op_block.operations[_ - 1].operation
+        current_last_pass = named_sequence_op_block.operations[_ - 2].operation
+        with ir.InsertionPoint(yield_op):
+            apply_registered_pass_op = ApplyRegisteredPassOp(
+                result=transform_func_type,
+                target=current_last_pass.result,
+                pass_name=pass_name,
+                options=options,
+            )
+
+    # otherwise it's the first pass, i.e. only a yield op is in the block
+    # so insert right before the yield op
+    elif first_op_in_block.name == "transform.yield":
+        ip = named_sequence_op.regions[0].blocks[0]
+        with ir.InsertionPoint(ip.operations[len(ip.operations) - 1]):
+            apply_registered_pass_op = ApplyRegisteredPassOp(
+                result=transform_func_type,
+                target=ip.arguments[0],
+                pass_name=pass_name,
+                options=options,
+            )
+
+    else:
+        raise RuntimeError(
+            """
+            Unexpected operation in transform.named_sequence! 
+            Only transform.apply_registered_pass and transform.yield are allowed.
+            """
+        )
```

Comment:
I would make this just an assertion and remove the else.

Reply:
done